src/WebSocketStream.cpp

```diff
@@ -173,15 +173,35 @@ namespace aws {
                 }
             }
 
-            void WebSocketStream::async_ssl_handshake(const ssl::stream_base::handshake_type &type,
+            void WebSocketStream::async_ssl_handshake(const ssl::stream_base::handshake_type &type, const std::string &host,
                                                       const BoostCallbackFunc &handler) {
                 if (localproxyConfig.is_web_proxy_using_tls) {
                     BOOST_LOG_SEV(*log, trace) << "Calling next_layer().async_handshake with type: "
                                                << WEB_PROXY_WITH_TLS_TYPE_NAME;
+                    // Set SNI Hostname (many hosts need this to handshake successfully)
+                    if(!SSL_set_tlsext_host_name(boost::get<unique_ptr<WEB_PROXY_WITH_TLS_TYPE>>(wss)->next_layer().native_handle(), host.c_str()))
+                    {
+                         BOOST_LOG_SEV(*log, trace) << "SSL next_layer() failed to set SNI";
+                    }
+                    else
+                    {
+                         BOOST_LOG_SEV(*log, trace) << "SSL next_layer() SNI is set : "
+                                                    << host;
+                    }
                     return boost::get<unique_ptr<WEB_PROXY_WITH_TLS_TYPE>>(wss)->next_layer().async_handshake(type, handler);
                 } else {
                     BOOST_LOG_SEV(*log, trace) << "Calling next_layer().async_handshake with type: "
                                                << WEB_PROXY_NO_TLS_TYPE_NAME;
+                    // Set SNI Hostname (many hosts need this to handshake successfully)
+                    if(!SSL_set_tlsext_host_name(boost::get<unique_ptr<WEB_PROXY_NO_TLS_TYPE>>(wss)->next_layer().native_handle(), host.c_str()))
+                    {
```

Comment:
Since this block is executed when the web proxy is not using TLS, Is it needed to set the SNI in this code branch/path?

Reply:
Hey @jaiswal-sumit , can you please address this and other comments raised by the team? Thank you

Reply:
Hello @HarshGandhi-AWS @kareali Its fine to set SNI even during web proxy connection calls too.
Below are the debug logs: As example: I have use regional aws default tunnel, in default behaviour with web proxy setting. Setting SNI is need in actual request which even go via Web proxy as server needs SNI information. I am open to discuss more on it, if it's needed. Thank you :)  

```
[2023-10-11 08:20:20.541774] (0x00007ff849dfd700) [trace] Establishing TCP connection with the Web Proxy
[2023-10-11 08:20:20.542750] (0x00007ff849dfd700) [debug] Connected successfully with Web Proxy
[2023-10-11 08:20:20.542771] (0x00007ff849dfd700) [trace] Calling lowest_layer with type: websocket_stream_single_ssl_type
[2023-10-11 08:20:20.542789] (0x00007ff849dfd700) [trace] Preparing HTTP CONNECT request
[2023-10-11 08:20:20.542800] (0x00007ff849dfd700) [trace] Sending HTTP CONNECT
[2023-10-11 08:20:20.542958] (0x00007ff849dfd700) [debug] Successfully sent HTTP CONNECT to the Web proxy
[2023-10-11 08:20:20.542990] (0x00007ff849dfd700) [trace] Waiting for HTTP CONNECT response from the Web proxy
[2023-10-11 08:20:20.811438] (0x00007ff849dfd700) [trace] Parsing the HTTPS response from the Web proxy

[2023-10-11 08:20:20.832312] (0x00007ff849dfd700) [info] TCP tunnel established successfully
[2023-10-11 08:20:20.832341] (0x00007ff849dfd700) [debug] Connected successfully with proxy server
[2023-10-11 08:20:20.832522] (0x00007ff849dfd700) [trace] Calling lowest_layer with type: websocket_stream_single_ssl_type
[2023-10-11 08:20:20.832544] (0x00007ff849dfd700) [trace] Calling lowest_layer with type: websocket_stream_single_ssl_type
[2023-10-11 08:20:20.832555] (0x00007ff849dfd700) [trace] Performing SSL handshake with proxy server
[2023-10-11 08:20:20.832565] (0x00007ff849dfd700) [trace] Calling set_verify_mode with type: single_ssl_stream
[2023-10-11 08:20:20.832574] (0x00007ff849dfd700) [trace] Calling set_verify_callback with type: single_ssl_stream
[2023-10-11 08:20:20.832586] (0x00007ff849dfd700) [trace] Calling next_layer().async_handshake with type: websocket_stream_single_ssl_type
**[2023-10-11 08:20:20.832596] (0x00007ff849dfd700) [trace] SSL next_layer() SNI is set : data.tunneling.iot.eu-west-1.amazonaws.com**
[2023-10-11 08:20:21.219403] (0x00007ff849dfd700) [debug] Successfully completed SSL handshake with proxy server
[2023-10-11 08:20:21.219451] (0x00007ff849dfd700) [trace] Performing websocket handshake with proxy server
[2023-10-11 08:20:21.219473] (0x00007ff849dfd700) [trace] Calling async_handshake with type: websocket_stream_single_ssl_type
[2023-10-11 08:20:21.219536] (0x00007ff849dfd700) [trace] Web socket ugprade request(*not entirely final):


[2023-10-11 08:20:21.443448] (0x00007ff849dfd700) [debug] Web socket subprotocol selected: aws.iot.securetunneling-3.0
[2023-10-11 08:20:21.443465] (0x00007ff849dfd700) [info] Successfully established websocket connection with proxy server: wss://data.tunneling.iot.eu-west-1.amazonaws.com:443
```